### PR TITLE
Use Abstract* base classes consistently to save on classfile size

### PR DESCRIPTION
**Pull request by lrytz**

### CODE REVIEW

**Comment by lrytz** [comment]
Before:

```
$ du -h build/quick/classes/library/scala/collection
2.0M	build/quick/classes/library/scala/collection/mutable
2.9M	build/quick/classes/library/scala/collection/immutable
248K	build/quick/classes/library/scala/collection/concurrent
104K	build/quick/classes/library/scala/collection/generic
532K	build/quick/classes/library/scala/collection/convert
9.7M	build/quick/classes/library/scala/collection
```

After:
```
$ du -h build/quick/classes/library/scala/collection
1.5M	build/quick/classes/library/scala/collection/mutable
1.9M	build/quick/classes/library/scala/collection/immutable
212K	build/quick/classes/library/scala/collection/concurrent
104K	build/quick/classes/library/scala/collection/generic
532K	build/quick/classes/library/scala/collection/convert
7.0M	build/quick/classes/library/scala/collection
```

File: src/library/scala/collection/immutable/BitSet.scala [hunk 1]
```diff
@@ -16,7 +16,8 @@ import mutable.{Builder, GrowableBuilder}
   */
 @SerialVersionUID(3L)
 sealed abstract class BitSet
-  extends SortedSet[Int]
+  extends AbstractSet[Int]
+    with SortedSet[Int]
```
Comment (julienrf) [inline src/library/scala/collection/immutable/BitSet.scala:20]: Why not have `AbstractSortedSet`?

File: src/library/scala/collection/immutable/Seq.scala [hunk 2]
```diff
@@ -74,3 +74,5 @@ object LinearSeq extends SeqFactory.Delegate[LinearSeq](List)
 trait LinearSeqOps[+A, +CC[X] <: LinearSeq[X], +C <: LinearSeq[A] with LinearSeqOps[A, CC, C]]
   extends Any with SeqOps[A, CC, C]
     with collection.LinearSeqOps[A, CC, C]
+
+abstract class AbstractSeq[+A] extends scala.collection.AbstractSeq[A] with Seq[A]
```
Comment (julienrf) [inline src/library/scala/collection/immutable/Seq.scala]: Do we also want `AbstractLinearSeq` or `AbstractIndexedSeq`? What would be the gain of having them?

**Review by julienrf — APPROVED** [review]
Some `AbstractXxx` types have no scaladoc, do you mind systematically adding them? If find the `Abstract` prefix not meaningful at all, so I think it’s good if we always document these types.

~~~ scala
 /** Explicit instantiation of the `Xxx` trait to reduce class file size in subclasses. */
~~~

**Comment by lrytz** [comment]
Hmm, getting VerifyErrors with bridges...
```
Exception in thread "main" java.lang.VerifyError: class scala.collection.immutable.List overrides final method toSeq.()Lscala/collection/Seq;
```
I'm trying to build a minimized example, haven't managed yet.

**Comment by julienrf** [comment]
IIRC @szeiger encountered the same error in the past. The workaround consisted in making the implementation non final...

**Comment by lrytz** [comment]
I certainly want to minimize this bug!

**Comment by lrytz** [comment]
Finally managed to minimize it: https://github.com/scala/bug/issues/10853. That took a long time..

File: src/library/scala/collection/immutable/BitSet.scala (hunk 1 above)
Comment (lrytz) [reply src/library/scala/collection/immutable/BitSet.scala:20]: It didn't seem worth it, as there are only two new methods in `SortedSet`

File: src/library/scala/collection/immutable/Seq.scala (hunk 2 above)
Comment (lrytz) [reply src/library/scala/collection/immutable/Seq.scala]: Again, there are few (none?) new methods in these traits, so they should not cause additional forwarders to be emitted.

Comment (lrytz) [reply src/library/scala/collection/immutable/Seq.scala]: Actually, I'm probably wrong here; for example `IndexedSeqOps` overrides a lot of methods, causes additional mixin forwarders. I'll take a look if it's worth adding more abstract classes.

File: src/library/scala/collection/LowPriority.scala [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package scala.collection
+
+class LowPriority {
```
Comment (szeiger) [inline src/library/scala/collection/LowPriority.scala:3]: Most (all?) package object contents including this `LowPriority` should be removed. They were used for implicit conversions in `strawman.collection` but are no longer necessary in `scala.collection` where `Predef` is used instead.

**Comment by szeiger** [comment]
Merged via https://github.com/scala/scala/pull/6594

**Pull request closed without merging**